File: android/app/src/main/java/enzio/android/eatfit/data/remote/GrpcAuthService.kt
package enzio.android.eatfit.data.remote

import enzio.android.eatfit.data.model.AuthResult
import enzio.android.eatfit.data.model.LoginRequest
import enzio.android.eatfit.data.model.RegisterRequest
import enzio.android.eatfit.proto.AuthServiceGrpc
import enzio.android.eatfit.proto.ForgotPasswordRequest
import enzio.android.eatfit.proto.LogoutRequest
import enzio.android.eatfit.proto.RegisterRequest as ProtoRegisterRequest
import enzio.android.eatfit.proto.ResetPasswordRequest
import enzio.android.eatfit.proto.VerifyEmailRequest
import enzio.android.eatfit.proto.LoginRequest as ProtoLoginRequest
import enzio.android.eatfit.proto.ChangeLoginRequest
import enzio.android.eatfit.proto.ChangePasswordRequest
import io.grpc.ManagedChannel
import io.grpc.okhttp.OkHttpChannelBuilder
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import java.util.concurrent.TimeUnit


class GrpcAuthService(
    private val serverUrl: String = "192.168.0.10",
    private val port: Int = 50051
) {

    private val channel: ManagedChannel = OkHttpChannelBuilder
        .forAddress(serverUrl, port)
        .usePlaintext()
        .keepAliveTime(30, TimeUnit.SECONDS)
        .keepAliveTimeout(10, TimeUnit.SECONDS)
        .keepAliveWithoutCalls(true)
        .build()

    private val stub = AuthServiceGrpc.newBlockingStub(channel)

    suspend fun register(request: RegisterRequest): AuthResult = withContext(Dispatchers.IO) {
        try {
            val protoRequest = ProtoRegisterRequest.newBuilder()
                .setEmail(request.email)
                .setLogin(request.login)
                .setPassword(request.password)
                .setPasswordConfirmation(request.passwordConfirmation)
                .build()

            val response = stub.register(protoRequest)

            AuthResult(
                success = response.success,
                message = response.message,
                userId = response.userId
            )
        } catch (e: Exception) {
            AuthResult(success = false, message = "Error: ${e.message}")
        }
    }

    suspend fun verifyEmail(userId: String, code: String): AuthResult = withContext(Dispatchers.IO) {
        try {
            val request = VerifyEmailRequest.newBuilder()
                .setUserId(userId)
                .setVerificationCode(code)
                .build()

            val response = stub.verifyEmail(request)

            AuthResult(
                success = response.success,
                message = response.message
            )
        } catch (e: Exception) {
            AuthResult(success = false, message = "Error: ${e.message}")
        }
    }

    suspend fun login(request: LoginRequest): AuthResult = withContext(Dispatchers.IO) {
        try {
            val builder = ProtoLoginRequest.newBuilder()
                .setPassword(request.password)
                .setDeviceInfo("Android Emulator")

            if (request.loginIdentifier.contains('@')) {
                builder.email = request.loginIdentifier
            } else {
                builder.login = request.loginIdentifier
            }

            val response = stub
                .withDeadlineAfter(10, TimeUnit.SECONDS)
                .login(builder.build())

            AuthResult(
                success = true,
                accessToken = response.accessToken,
                refreshToken = response.refreshToken,
                userId = response.user.userId,
                email = response.user.email,
                login = response.user.login
            )
        } catch (e: Exception) {
            AuthResult(success = false, message = "Login failed: ${e.message}")
        }
    }

    suspend fun forgotPassword(loginIdentifier: String): AuthResult = withContext(Dispatchers.IO) {
        try {
            val request = ForgotPasswordRequest.newBuilder()
                .setLoginIdentifier(loginIdentifier)
                .build()

            val response = stub.forgotPassword(request)

            AuthResult(
                success = response.success,
                message = response.message,
                resetToken = response.resetToken
            )
        } catch (e: Exception) {
            AuthResult(success = false, message = "Error: ${e.message}")
        }
    }

    suspend fun resetPassword(resetToken: String, code: String, newPassword: String): AuthResult = withContext(Dispatchers.IO) {
        try {
            val request = ResetPasswordRequest.newBuilder()
                .setResetToken(resetToken)
                .setVerificationCode(code)
                .setNewPassword(newPassword)
                .setPasswordConfirmation(newPassword)
                .build()

            val response = stub.resetPassword(request)

            AuthResult(
                success = response.success,
                message = response.message
            )
        } catch (e: Exception) {
            AuthResult(success = false, message = "Error: ${e.message}")
        }
    }

    suspend fun logout(refreshToken: String) {
        withContext(Dispatchers.IO) {
            try {
                val request = LogoutRequest.newBuilder()
                    .setRefreshToken(refreshToken)
                    .build()
                stub
                    .withDeadlineAfter(5, java.util.concurrent.TimeUnit.SECONDS)
                    .logout(request)
            } catch (_: Exception) {
                // Игнорируем ошибки
            }
        }
    }

    suspend fun changeLogin(userId: String, newLogin: String, password: String): AuthResult = withContext(Dispatchers.IO) {
        try {
            val request = ChangeLoginRequest.newBuilder()
                .setUserId(userId)
                .setNewLogin(newLogin)
                .setPassword(password)
                .build()
            val response = stub.changeLogin(request)
            AuthResult(success = response.success, message = response.message)
        } catch (e: Exception) {
            AuthResult(success = false, message = e.message ?: "Error")
        }
    }

    suspend fun changePassword(userId: String, currentPassword: String, newPassword: String, confirmPassword: String): AuthResult = withContext(Dispatchers.IO) {
        try {
            val request = ChangePasswordRequest.newBuilder()
                .setUserId(userId)
                .setCurrentPassword(currentPassword)
                .setNewPassword(newPassword)
                .setPasswordConfirmation(confirmPassword)
                .build()
            val response = stub.changePassword(request)
            AuthResult(success = response.success, message = response.message)
        } catch (e: Exception) {
            AuthResult(success = false, message = e.message ?: "Error")
        }
    }
}

File: desktop/EatfitDesktop/Services/GrpcAuthService.cs
using System;
using System.Net.Http;
using System.Threading.Tasks;
using Eatfit.V1;
using EatfitDesktop.Models;
using EatfitDesktop.Services.Interfaces;
using Grpc.Core;
using Grpc.Net.Client;

namespace EatfitDesktop.Services
{
    public class GrpcAuthService : IAuthService, IDisposable
    {
        private readonly GrpcChannel _channel;
        private readonly Eatfit.V1.AuthService.AuthServiceClient _client;
        private bool _disposed;

        public GrpcAuthService(string serverUrl = "http://localhost:50051")
        {
            var handler = new HttpClientHandler
            {
                ServerCertificateCustomValidationCallback =
                    HttpClientHandler.DangerousAcceptAnyServerCertificateValidator
            };

            _channel = GrpcChannel.ForAddress(serverUrl, new GrpcChannelOptions
            {
                HttpHandler = handler
            });

            _client = new Eatfit.V1.AuthService.AuthServiceClient(_channel);
        }

        public async Task<AuthResult> RegisterAsync(RegisterModel model)
        {
            try
            {
                var request = new RegisterRequest
                {
                    Email = model.Email,
                    Login = model.Login,
                    Password = model.Password,
                    PasswordConfirmation = model.PasswordConfirmation
                };

                var response = await _client.RegisterAsync(request);

                return new AuthResult
                {
                    Success = response.Success,
                    Message = response.Message,
                    UserId = response.UserId
                };
            }
            catch (RpcException ex)
            {
                return new AuthResult { Success = false, Message = ex.Status.Detail };
            }
        }

        public async Task<AuthResult> VerifyEmailAsync(string userId, string code)
        {
            try
            {
                var request = new VerifyEmailRequest
                {
                    UserId = userId,
                    VerificationCode = code
                };

                var response = await _client.VerifyEmailAsync(request);

                return new AuthResult
                {
                    Success = response.Success,
                    Message = response.Message
                };
            }
            catch (RpcException ex)
            {
                return new AuthResult { Success = false, Message = ex.Status.Detail };
            }
        }

        public async Task<AuthResult> LoginAsync(LoginModel model)
        {
            try
            {
                var request = new LoginRequest
                {
                    Password = model.Password,
                    DeviceInfo = $"WPF Desktop - {Environment.MachineName}"
                };

                if (model.LoginIdentifier.Contains('@'))
                    request.Email = model.LoginIdentifier;
                else
                    request.Login = model.LoginIdentifier;

                var response = await _client.LoginAsync(request);

                return new AuthResult
                {
                    Success = true,
                    Message = "Login successful",
                    AccessToken = response.AccessToken,
                    RefreshToken = response.RefreshToken,
                    UserId = response.User.UserId,
                    Email = response.User.Email,
                    Login = response.User.Login
                };
            }
            catch (RpcException ex)
            {
                return new AuthResult { Success = false, Message = ex.Status.Detail };
            }
        }

        public async Task<AuthResult> ForgotPasswordAsync(string loginIdentifier)
        {
            try
            {
                var request = new ForgotPasswordRequest
                {
                    LoginIdentifier = loginIdentifier
                };

                var response = await _client.ForgotPasswordAsync(request);

                return new AuthResult
                {
                    Success = response.Success,
                    Message = response.Message,
                    UserId = response.ResetToken // Используем как временное хранилище токена
                };
            }
            catch (RpcException ex)
            {
                return new AuthResult { Success = false, Message = ex.Status.Detail };
            }
        }

        public async Task<AuthResult> ResetPasswordAsync(string resetToken, string code, string newPassword)
        {
            try
            {
                var request = new ResetPasswordRequest
                {
                    ResetToken = resetToken,
                    VerificationCode = code,
                    NewPassword = newPassword,
                    PasswordConfirmation = newPassword // Если подтверждение не нужно отдельно
                };

                var response = await _client.ResetPasswordAsync(request);

                return new AuthResult
                {
                    Success = response.Success,
                    Message = response.Message
                };
            }
            catch (RpcException ex)
            {
                return new AuthResult { Success = false, Message = ex.Status.Detail };
            }
        }

        public async Task<AuthResult> DeleteAccountAsync(string userId)
        {
            await Task.CompletedTask;
            return new AuthResult { Success = true };
        }

        public async Task<AuthResult> RefreshTokenAsync(string refreshToken)
        {
            try
            {
                var request = new RefreshTokenRequest { RefreshToken = refreshToken };
                var response = await _client.RefreshTokenAsync(request);

                return new AuthResult
                {
                    Success = true,
                    AccessToken = response.AccessToken,
                    RefreshToken = response.RefreshToken
                };
            }
            catch
            {
                return new AuthResult { Success = false, Message = "Session expired" };
            }
        }

        public async Task LogoutAsync(string refreshToken)
        {
            try
            {
                await _client.LogoutAsync(new LogoutRequest { RefreshToken = refreshToken });
            }
            catch
            {
                // Игнорируем ошибки при выходе
            }
        }

        public async Task<AuthResult> ChangeLoginAsync(string userId, string newLogin, string password)
        {
            try
            {
                var request = new ChangeLoginRequest
                {
                    UserId = userId,
                    NewLogin = newLogin,
                    Password = password
                };
                var response = await _client.ChangeLoginAsync(request);
                return new AuthResult { Success = response.Success, Message = response.Message };
            }
            catch (RpcException ex)
            {
                return new AuthResult { Success = false, Message = ex.Status.Detail };
            }
        }

        public async Task<AuthResult> ChangePasswordAsync(string userId, string currentPassword,
            string newPassword, string confirmPassword)
        {
            try
            {
                var request = new ChangePasswordRequest
                {
                    UserId = userId,
                    CurrentPassword = currentPassword,
                    NewPassword = newPassword,
                    PasswordConfirmation = confirmPassword
                };
                var response = await _client.ChangePasswordAsync(request);
                return new AuthResult { Success = response.Success, Message = response.Message };
            }
            catch (RpcException ex)
            {
                return new AuthResult { Success = false, Message = ex.Status.Detail };
            }
        }

        public void Dispose()
        {
            if (_disposed) return;
            _disposed = true;
            _channel?.Dispose();
            GC.SuppressFinalize(this);
        }
    }
}